include/plugin_loader/plugin_loader.hpp

```diff
@@ -0,0 +1,242 @@
+        boost::dll::load_mode::append_decorations | boost::dll::load_mode::search_system_folders;
+    lib = boost::dll::shared_library(sl, ec, mode);
+  }
+  else
+  {
+    boost::filesystem::path sl = boost::filesystem::path(library_directory) / library_name;
+    lib = boost::dll::shared_library(sl, ec, boost::dll::load_mode::append_decorations);
+  }
+
+  // Check if it failed to find or load library
+  if (ec)
+  {
+    std::cerr << "Failed to find or load library: " << decorate(library_name, library_directory).c_str()
+              << " with error: " << ec.message();
+    return false;
+  }
+
+  return lib.has(symbol_name);
+}
+
+std::set<std::string> parseEnvironmentVariableList(const std::string& env_variable)
+{
+  std::set<std::string> list;
+  char* env_var = std::getenv(env_variable.c_str());
+  if (env_var == nullptr)  // Environment variable not found
+    return list;
+
+  std::string evn_str = std::string(env_var);
+  boost::split(list, evn_str, boost::is_any_of(";"), boost::token_compress_on);
+  return list;
+}
+
+std::set<std::string> getAllSearchPaths(const std::string& search_paths_env,
+                                        const std::set<std::string>& existing_search_paths)
+{
+  // Check for environment variable to override default library
+  if (!search_paths_env.empty())
+  {
+    std::set<std::string> search_paths = parseEnvironmentVariableList(search_paths_env);
+    search_paths.insert(existing_search_paths.begin(), existing_search_paths.end());
+    return search_paths;
+  }
+
+  return existing_search_paths;
+}
+
+std::set<std::string> getAllLibraryNames(const std::string& search_libraries_env,
+                                         const std::set<std::string>& existing_search_libraries)
+{
+  // Check for environment variable to override default library
+  if (!search_libraries_env.empty())
+  {
+    std::set<std::string> search_libraries = parseEnvironmentVariableList(search_libraries_env);
+    search_libraries.insert(existing_search_libraries.begin(), existing_search_libraries.end());
+    return search_libraries;
+  }
+
+  return existing_search_libraries;
+}
+
+boost::dll::shared_library loadLibrary(const std::string& library_name, const std::string& library_directory = "")
+{
+  boost::system::error_code ec;
+  boost::dll::shared_library lib;
+  if (library_directory.empty())
+  {
+    boost::filesystem::path sl(library_name);
+    boost::dll::load_mode::type mode =
+        boost::dll::load_mode::append_decorations | boost::dll::load_mode::search_system_folders;
+    lib = boost::dll::shared_library(sl, ec, mode);
+  }
+  else
+  {
+    boost::filesystem::path sl = boost::filesystem::path(library_directory) / library_name;
+    lib = boost::dll::shared_library(sl, ec, boost::dll::load_mode::append_decorations);
+  }
+
+  // Check if it failed to find or load library
+  if (ec)
+    throw std::runtime_error("Failed to find or load library: " + decorate(library_name, library_directory) +
```

Comment:
Might consider throwing a more specific type.

Reply:
Like a custom type? I don't see any other [out-of-the box exceptions](https://en.cppreference.com/w/cpp/error/exception) that fit much better

Reply:
Yes, I meant creating a custom type.